### PR TITLE
optimizer: code quality and performance improvements to post-opt analysis

### PR DESCRIPTION
**Pull request by aviatesk**
This PR is composed of a set of fixes and improvements to `ipo_dataflow_analysis!`.
I will rebase this once we merge #51185 and #51188, so still WIP, but let's see benchmark.

@nanosoldier `runbenchmarks("inference", vs=":master")`

### CODE REVIEW

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51211) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/6d61d2e_vs_78e0ca3/report.md).

**Comment by aviatesk** [comment]
@nanosoldier `runbenchmarks("inference", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51211#issuecomment-1707787090) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/cbfa4d3_vs_f066500/report.md).

File: base/compiler/ssair/irinterp.jl [hunk 1]
```diff
@@ -217,7 +217,7 @@ function BBScanner(ir::IRCode)
     return BBScanner(ir, bb_ip)
 end
 
-function scan!(callback, scanner::BBScanner, forwards_only::Bool)
+function scan!(@specialize(callback), scanner::BBScanner, forwards_only::Bool)
```
Comment (Keno) [inline base/compiler/ssair/irinterp.jl:220]: Is this a good idea? The inner loop seems potentially hot.

**Comment by aviatesk** [comment]
@nanosoldier `runbenchmarks("inference", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51211#issuecomment-1711158057) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/049a03d_vs_de297fa/report.md).

**Comment by aviatesk** [comment]
@nanosoldier `runbenchmarks("inference", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51211#issuecomment-1711318591) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/330343c_vs_de297fa/report.md).

**Comment by aviatesk** [comment]
@nanosoldier `runbenchmarks("inference")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51211#issuecomment-1711630938) has completed - successfully executed benchmarks. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/617f776/report.md).

**Comment by vtjnash** [comment]
@nanosoldier `runbenchmarks("inference", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51211#issuecomment-1711845834) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/12a9bac_vs_8e77b63/report.md).

**Comment by aviatesk** [comment]
Locally I confirmed performance improvements by running `inference` benchmarks against the latest mater, but somehow they aren't reproduced on nanosoldier..

**Comment by aviatesk** [comment]
@nanosoldier `runbenchmarks("inference", vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/51211#issuecomment-1711953783) has completed - no performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/NanosoldierReports/blob/master/benchmark/by_hash/5475124_vs_8e77b63/report.md).

**Pull request merged**